src/vmm/src/builder.rs

```diff
@@ -781,31 +772,23 @@ fn attach_net_devices(
     Ok(())
 }
 
-fn attach_vsock_device(
+fn attach_unixsock_vsock_device(
     vmm: &mut Vmm,
-    vsock: &VsockDeviceConfig,
+    unix_vsock: &Arc<Mutex<Vsock<VsockUnixBackend>>>,
     event_manager: &mut EventManager,
 ) -> std::result::Result<(), StartMicrovmError> {
     use self::StartMicrovmError::*;
-    let backend = devices::virtio::vsock::VsockUnixBackend::new(
-        u64::from(vsock.guest_cid),
-        vsock.uds_path.clone(),
-    )
-    .map_err(CreateVsockBackend)?;
-
-    let vsock_device = Arc::new(Mutex::new(
-        devices::virtio::Vsock::new(u64::from(vsock.guest_cid), backend)
-            .map_err(CreateVsockDevice)?,
-    ));
 
     event_manager
-        .add_subscriber(vsock_device.clone())
-        .map_err(StartMicrovmError::RegisterEvent)?;
+        .add_subscriber(unix_vsock.clone())
+        .map_err(RegisterEvent)?;
 
+    let id = String::from(unix_vsock.lock().unwrap().id());
+    // The device mutex mustn't be locked here otherwise it will deadlock.
     attach_mmio_device(
         vmm,
-        vsock.vsock_id.clone(),
-        MmioTransport::new(vmm.guest_memory().clone(), vsock_device)
+        id,
```

Comment:
This cannot be removed as `Block` still uses it. Only the last merging PR can remove it (this or #1736 whichever is last to go in).

Furthermore, after looking a bit at the code implications I don't know if it's worth removing it.

Mmio device identification happens on a generic `VirtioDevice` based on tuple `(dev.type_id(), dev_id)` where `dev_id` is string.
`type_id()` is part of the `VirtioDevice` trait inteface, but `id()` isn't and doesn't belong there either. So we can't remove it from the param while still using generic `VirtioDevice`s in the inner hashmap. To remove it we'd need to define a wrapper trait that also adds `id()` to the list of virtio methods and use that as a type in the hashmap.

Don't feel it's worth it anymore.

Reply:
Yeah I thought the Block PR would go in first so this PR would have needed to enclose all the modifications to remove `id` from all devices. I understand your point. The ID for the application that consumes a device is represented by that tuple. On its own (if you took the physical device out of the computer) it would still have an identifier (you can observe its type); that would be the `device_type` in the `VirtioDevice`, although a device doesn't necessarily need to be `virtio` to have an observable type :thinking: Anyway, putting the device back into the computer, its identity is extended, it's in a certain slot etc - we skip that; with the OS running, the device will have an identifier allocated. That's where the 2nd component comes in. So conceptually, the way I see it, the "id" is a layered thing composed of multiple identifiers that come from different facets of what a device represents. Put into Rust, it would be a tuple of ids coming from multiple traits. Not necessarily a wrapper, but maybe a `VirtioDevice + OtherDeviceTraitThatHasId`. I.e. thing that makes sense for a standalone device, plus thing that makes sense only when the device is plugged into Firecracker.

...That said, I can see how it's way out of scope for now. I still think it's the right thing to do, and I'd track it as a part of https://github.com/firecracker-microvm/firecracker/issues/1408. WDYT?

Reply:
Yup. Everything you said makes perfect sense.